File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-3/version-1/javascript/worker.js
// worker.js
importScripts("https://cdn.jsdelivr.net/npm/danfojs@1.1.2/lib/bundle.js");

self.onmessage = function (event) {
    const { id, size } = event.data;
    if (!size) {
        self.postMessage({ id, error: "Size not provided" });
        return;
    }

    try {
        const metrics = {};
        const startTotal = performance.now();

        const startCreate = performance.now();
        const dataArray = Array.from({ length: size }, () => {
            return [Math.floor(Math.random() * 1000)];
        });
        const df = new self.dfd.DataFrame(dataArray, { columns: ["values"] });
        const memoryUsage = (size * 4) / (1024 * 1024);
        metrics.create = {
            time: performance.now() - startCreate,
            memory: memoryUsage,
        };

        const sumStart = performance.now();
        const sumValue = df.sum();
        metrics.sum = {
            time: performance.now() - sumStart,
            memory: memoryUsage,
        };

        const meanStart = performance.now();
        const meanValue = df.mean();
        metrics.mean = {
            time: performance.now() - meanStart,
            memory: memoryUsage,
        };

        const stdStart = performance.now();
        const stdValue = df.std();
        metrics.std = {
            time: performance.now() - stdStart,
            memory: memoryUsage,
        };

        metrics.total = { time: performance.now() - startTotal };
        self.postMessage({ id, results: metrics });
    } catch (e) {
        console.error("Error:", e);
        self.postMessage({
            id,
            error: e.message,
            stack: e.stack,
        });
    }
};

File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-3/version-1/python/worker.py
import time
import numpy as np
import pandas as pd
from pyscript import sync
import json


seed = int(time.time() * 1000) % 2**32
rng = np.random.default_rng(seed)


def do_statistical_analysis(size):
    metrics = {
        "create": {"time": 0.0, "memory": 0.0},
        "sum":    {"time": 0.0, "memory": 0.0},
        "mean":   {"time": 0.0, "memory": 0.0},
        "std":    {"time": 0.0, "memory": 0.0},
        "total":  {"time": 0.0},
        "error":  None,
    }

    try:
        total_start = time.perf_counter()

        start = time.perf_counter()
        data = pd.DataFrame(
            rng.integers(0, 1000, size=(size, 1), dtype=np.int32),
            columns=["values"]
        )
        metrics["create"]["time"] = (time.perf_counter() - start) * 1000
        metrics["create"]["memory"] = data.memory_usage(
            deep=True).sum() / (1024 ** 2)

        for op in ("sum", "mean", "std"):
            metrics[op]["memory"] = metrics["create"]["memory"]

        start = time.perf_counter()
        _ = data["values"].sum()
        metrics["sum"]["time"] = (time.perf_counter() - start) * 1000

        start = time.perf_counter()
        _ = data["values"].mean()
        metrics["mean"]["time"] = (time.perf_counter() - start) * 1000

        start = time.perf_counter()
        _ = data["values"].std(ddof=0)
        metrics["std"]["time"] = (time.perf_counter() - start) * 1000

        metrics["total"]["time"] = (time.perf_counter() - total_start) * 1000

    except Exception as e:
        metrics["error"] = str(e)

    return json.dumps(metrics)


sync.do_statistical_analysis = do_statistical_analysis
